### ISSUE TITLE
await parent() causes parent layout load to re-run

### ISSUE THREAD

**Issue by hyunbinseo** [sveltejs/kit#8579]
### Describe the bug

According to the [loading data](https://kit.svelte.dev/docs/load) documentation,

> To summarize, a `load` function will re-run in the following situations:
> - It calls `await parent()` and a parent `load` function re-ran 

A child `load` function that calls `await parent()` should not cause parent `load` function(s) to re-run.

It is the other way around - the child `load` function should re-run if the parent `load` function **re-ran**.

However, calling `await parent()` seems to cause the parent `load` function(s) to re-run.

### Reproduction

Steps
1. Open the [StackBlitz](https://stackblitz.com/edit/sveltejs-kit-template-default-2ycnsp) project.
2. Alternately click `veritas` and `lacetti` links.
3. Check the console.

Logs
```bash
/routes/[sedan]/[id]/+layout.server.ts
/routes/[sedan]/[veritas]/+layout.server.ts

/routes/[sedan]/[lacetti]/+layout.server.ts
/routes/[sedan]/[id]/+layout.server.ts # not expected

/routes/[sedan]/[veritas]/+layout.server.ts
/routes/[sedan]/[id]/+layout.server.ts # not expected
```

---

Switching between `/sedan/veritas` and `/sedan/lacetti` should not cause the following function to re-run.

```javascript
/** @type {import('./$types').LayoutServerLoad} */
export async function load({ params: { type } }) {
  console.log(`/routes/[${type}]/[id]/+layout.server.ts`);
  return {};
}
```

It should re-run only if the `params.type` has changed. It has not. The value stays as `sedan`.

---

Fix

1. Remove `await parent();` from `/routes/[type]/[id]/+layout.server.ts`
2. Alternately click `veritas` and `lacetti` links.
3. Check the console.

Logs
```
/routes/[sedan]/[id]/+layout.server.ts
/routes/[sedan]/[veritas]/+layout.server.ts

/routes/[sedan]/[lacetti]/+layout.server.ts

/routes/[sedan]/[veritas]/+layout.server.ts
```

### System Info

```Shell
❯ npm list
├─ @sveltejs/adapter-auto@1.0.1
├─ @sveltejs/kit@1.1.1
├─ @sveltejs/vite-plugin-svelte@2.0.2
├─ svelte@3.55.1
├─ svelte-check@3.0.2
└─ vite@4.0.4
```

### Severity

annoyance

**Comment by dummdidumm** [comment]
Giving this the documentation because this works as designed. If you are using `await parent()` in a server `load` function, that parent data needs to be made available somehow, and passing the data back to the server would not scale for larger payloads. Since the server is stateless, the parent `load` function needs to rerun in order to make that data available.

**Comment by osmanlodzero** [comment]
I have a similar issue, but with client loads. 

From a childs load function, I'm calling await parent() and then throwing a redirect. This causes the parent load to be called twice. The reason for the redirect is to automatically select a child route ie: go from /foo/ to /foo/bar 

The parent load calls a slow API so I'd like to avoid that. 

Is this intended, if so what is a better way to load a list of data and then automatically select the first one AND change the route accordingly? Thanks!

**Comment by gterras** [comment]
> Giving this the documentation because this works as designed. If you are using `await parent()` in a server `load` function, that parent data needs to be made available somehow, and passing the data back to the server would not scale for larger payloads. Since the server is stateless, the parent `load` function needs to rerun in order to make that data available.

I understand this, but could this be avoided on initial load if no client load is involved?

Say I have a root `+layout.server` fetching data, and a root `+page.server` that is only meant to be a redirect to deeper levels (so there is no root `+page.svelte` file).  

This `+page.server` needs the layout data to generate the url it will `redirect()` to. The main goal of doing this in server is to allow SSRing the destination page on initial load (after that the root page is never reached again because the client has the data to generate the destination link when needed). 

This works as expected except the layout load is ran twice. It seems to me like it's avoidable in theory, but maybe I'm wrong or maybe it's too much of an edge case.

**Comment by eltigerchino** [comment]
This is already documented in https://svelte.dev/docs/kit/load#Rerunning-load-functions-When-do-load-functions-rerun

> ### When do load functions rerun?
> To summarize, a load function will rerun in the following situations:
> ...
> A child load function calls await parent() and is rerunning, and the parent is a server load function

**Comment by gterras** [comment]
> This is already documented

Yes but ain't the use case above legitimate? Should I create a dedicated issue?

**Comment by eltigerchino** [comment]
> > This is already documented
> 
> Yes but ain't the use case above legitimate? Should I create a dedicated issue?

> This works as expected except the layout load is ran twice on initial load. It seems to me like it's avoidable in theory, but maybe I'm wrong or maybe it's too much of an edge case.

This doesn't sound like the correct behaviour. The layout load should only run once if it doesn't have any dependencies. If this is a bug, you should create a new issue with a minimal reproduction.
